File: day02/gift_shop_sets.pl
#!/usr/bin/env perl
use v5.42;
use Set::Scalar;
use List::Util qw(sum);

my @ranges;
while (<<>>) {
    chomp;
    for my $pair (split ',') {
        push @ranges, [split '-', $pair];
    }
}

my %pairs = map {$_ x 2 => 1} 1..99_999;
my %dups = map {$_ x 2 => 1} 1..99_999;

$dups{$_ x 3} = 1 for 1..999;
for my $copies (4..5) {
    $dups{$_ x $copies} = 1 for 1..99;
}

for my $copies (6..10) {
    $dups{$_ x $copies} = 1 for 1..9;
}

my ($part1, $part2);
for my $range (@ranges) {
    for my $i ($range->[0]..$range->[1]) {
        $part1 += $i if defined $pairs{$i};
        $part2 += $i if defined $dups{$i};
    }
}

say "Part 1: $part1";
say "Part 2: $part2";

File: day02/gift_shop_sets.py
from sys import argv
from itertools import product

with open(argv[1]) as f:
    ranges = [[int(d) for d in p.split('-')] for p in [pair for pair in f.read().rstrip().split(',')]]

pairs = {int(str(x) * 2) for x in range(1, 100_000)}

dups = pairs.copy()
# 3 copies
dups |= {int(str(i) * 3) for i in range(1,1000)}

# 4-5 copies
dups |= {int(str(i) * copies) for i,copies in product(range(1,100), range(4,6))}

# 6-10 copies
dups |= {int(str(i) * copies) for i,copies in product(range(1,10), range(6,11))}

part1 = 0
part2 = 0
for r in ranges:
    part1 += sum(set(range(r[0], r[1]+1)) & pairs)
    part2 += sum(set(range(r[0], r[1]+1)) & dups)

print('Part 1:', part1)
print('Part 2:', part2)
